File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices):
        ans = prices[:]
        stack = []

        for i in range(len(prices)):

            while stack and prices[stack[-1]] >= prices[i]:

                index = stack.pop()

                ans[index] = prices[index] - prices[i]

            stack.append(i)

        return ans

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
import java.util.*;

class Solution {
    public int[] finalPrices(int[] prices) {

        int[] ans = prices.clone();

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < prices.length; i++) {

            while (!stack.isEmpty() &&
                   prices[stack.peek()] >= prices[i]) {

                int index = stack.pop();

                ans[index] = prices[index] - prices[i];
            }

            stack.push(i);
        }

        return ans;
    }
}